tests/classifiers/test_keras.py

```diff
@@ -397,12 +397,171 @@ def test_repr(self):
         self.assertIn('input_layer=0, output_layer=0', repr_)
 
     def test_loss_functions(self):
-        loss_names = ['categorical_hinge', 'categorical_crossentropy', 'sparse_categorical_crossentropy',
-                      'binary_crossentropy', 'kullback_leibler_divergence', 'cosine_proximity']
-
-        for loss_name in loss_names:
-            classifier = get_classifier_kr(loss_name=loss_name)
-            classifier.fit(x=self.x_test, y=self.y_test, nb_epochs=1)
-            classifier.predict(x=self.x_test)
-            classifier.class_gradient(self.x_test, label=5)
-            classifier.loss_gradient(x=self.x_test, y=self.y_test)
+
+        # prediction and class_gradient should be independent of logits/probabilities and of loss function
+
+        y_test_pred_expected = np.asarray([7, 1, 1, 4, 4, 1, 4, 4, 4, 4, 4, 4, 4, 4, 1, 4, 4, 7, 8, 4, 4, 4, 4, 3, 4, 4,
+                                           7, 4, 4, 4, 7, 4, 3, 4, 7, 0, 7, 7, 1, 1, 7, 7, 4, 0, 1, 4, 4, 4, 4, 4, 4, 7,
+                                           4, 4, 4, 4, 4, 1, 4, 4, 7, 4, 4, 4, 4, 4, 4, 4, 3, 0, 7, 4, 0, 1, 7, 4, 4, 7,
+                                           4, 4, 4, 4, 4, 7, 4, 4, 4, 4, 4, 1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4])
+
+        class_gradient_probabilities_expected = np.asarray([0.0000000e+00, 0.0000000e+00, 0.0000000e+00, 2.3582461e-03,
```

Comment:
The recent tendency has been to assert exact values in ART tests, but these are generally not very stable in the long run and require a lot of maintenance. Depending on the machine architecture and versions of packages, these numbers are not necessarily reproducible. E.g., in my case, versions of ART that pass on Travis fail locally. Should we think about another way of testing effectively?

Reply:
I agree that recent versions of ART had too much variability in some of these new unittests, but I think we have removed or improved most of these unstable tests on `dev_1.1.0`. The unstable testes were mostly for classes containing algorithms with a lot of randomness and initially we thought that random seeds would fix it. I like this kind of tests in cases where the results are reproducible, because they run fast and capture changing behaviour quite well. But I'm also open to other types of testing too.